Fix false-positive leak detection report when ReferenceCountedOpenSsl…

normanmaurer opened the pull request:
…Engine constructor throws.

 Motivation:

    We need to ensure we only create the ResourceLeak when the constructor not throws.

    Modifications:

    Ensure ResourceLeakDetector.track(...) is only called if the constructor of ReferenceCoundedOpenSslEngine not throws.

    Result:

    No more false-positves.

normanmaurer commented:
Notices this when doing something stupid and so had the constructor throw :)

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngineTest.java:
```diff
@@ -54,4 +56,14 @@ protected void cleanupServerSslContext(SslContext ctx) {
     }
+
+    @Test(expected = NullPointerException.class)
+    public void testNotLeakOnException() throws Exception {
```
this produced a leak and so would fail the build when run with `-Pleak` on the ci.

normanmaurer commented:
also @carl-mastrangelo

carl-mastrangelo commented on handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -236,56 +236,66 @@ protected void deallocate() {
         super(peerHost, peerPort);
         OpenSsl.ensureAvailability();
```
This could leak too, right?

normanmaurer replied:
the leak is created after this so it may be fine ?

carl-mastrangelo commented on handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -236,56 +236,66 @@ protected void deallocate() {
+            } finally {
+                readerLock.unlock();
+            }
+            synchronized (this) {
```
I think it would be safer if you just assigned `refCnt` here, after you have created `ssl`.   That would fix the problem of exceptions being thrown in initializers, or in super(), or in OpenSsl.ensureAvailability();

normanmaurer replied:
I thought about this but the subclass calls release() in the finalizer so I want to ensure everything is setup already.

Scottmitch commented on handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -236,56 +236,66 @@ protected void deallocate() {
         super(peerHost, peerPort);
         OpenSsl.ensureAvailability();
         leak = leakDetection ? leakDetector.track(this) : null;
```
can we just create this leak detector as the last line in the constructor (with a comment for justification)? This will also minimize the "risk" for letting `this` escaping the constructor and using unitized fields of this object (even though practically it shouldn't). The member variable is `final` so we shouldn't have to worry about re-ordering with instructions outside the scope of the constructor.

normanmaurer commented:
Yep that's a better idea... let me do this

> Am 29.07.2017 um 18:00 schrieb Scott Mitchell <notifications@github.com>:
> 
> @Scottmitch requested changes on this pull request.
> 
> In handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:
> 
> > @@ -236,56 +236,66 @@ protected void deallocate() {
>          super(peerHost, peerPort);
>          OpenSsl.ensureAvailability();
>          leak = leakDetection ? leakDetector.track(this) : null;
> can we just create this leak detector as the last line in the constructor (with a comment for justification)? This will also minimize the "risk" for letting this escaping the constructor and using unitized fields of this object (even though practically it shouldn't). The member variable is final so we shouldn't have to worry about re-ordering with instructions outside the scope of the constructor.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

normanmaurer commented:
Cherry-picked into 4.1 (c5b5d363601a0d9ae294034c1e68a6de54a0958f) and 4.0 (3a97fdb8be6d6011836c997ec48d1c4643e9391a) after addressing the comment of @Scottmitch

Scottmitch commented on handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -288,6 +287,7 @@ protected void deallocate() {
                 PlatformDependent.throwException(cause);
             }
         }
+        leak = leakDetection ? leakDetector.track(this) : null;
```
would be nice to add a comment explaining why this is at the end so we don't move it by accident later.

normanmaurer replied:
Done... d9d3d65716

normanmaurer commented:
Sure let me do as a followup commit tomorrow when I am back at my laptop

> Am 29.07.2017 um 23:28 schrieb Scott Mitchell <notifications@github.com>:
> 
> @Scottmitch commented on this pull request.
> 
> In handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:
> 
> > @@ -288,6 +287,7 @@ protected void deallocate() {
>                  PlatformDependent.throwException(cause);
>              }
>          }
> +        leak = leakDetection ? leakDetector.track(this) : null;
> would be nice to add a comment explaining why this is at the end so we don't move it by accident later.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
